Grails 6.2.3 - Cast Exception in Controller Action with a parameter with the same name as a local variable in a different action

### Expected Behavior

No incorrect cast should be performed on a controller action that takes a String parameter. 

### Actual Behaviour

In the example project attached to this ticket you can see that the `exampleController`  defines an `echo` action, this action takes a `String` parameter named `person`. When another action in the same controller defines a local variable with the same name (`person`) and a different type (type `Person` in the example), an incorrect cast in being performed. The `person` parameter, which should be received as `String` is being converted to the type `Person`  and a `GroovyCastException` is thrown.

This issue is similar to #13486


### Steps To Reproduce

Details in [README.md](https://github.com/solaechea/grails-broken-binding)

1. Launch grails app
1. Visit http://localhost:8080/example/echo?person=George+Doe

### Environment Information

Windows 11 Home
Version	24H2
OS build	26100.4652
Experience Windows Feature Experience Pack 1000.26100.128.0

java 17.0.9 2023-10-17 LTS
Java(TM) SE Runtime Environment (build 17.0.9+11-LTS-201)
Java HotSpot(TM) 64-Bit Server VM (build 17.0.9+11-LTS-201, mixed mode, sharing)


### Example Application

https://github.com/solaechea/grails-broken-binding

### Version

6.2.3

I've confirmed this is still an issue in Grails 7.  https://github.com/apache/grails-core/commit/b241c86429dc74159b219b87a8342a102b60735a adds a test to grails-core to reproduce this. 

Here's the full stacktrace: 


                2025-08-06 09:30:07.890 ERROR --- [nio-8080-exec-1] o.g.web.errors.GrailsExceptionResolver   : GroovyCastException occurred when processing request: [GET] /commandObject/echo - parameters:
                person: George Doe
                Cannot cast object 'George Doe' with class 'java.lang.String' to class 'functionaltests.Person'. Stacktrace follows:

                java.lang.reflect.InvocationTargetException: null
                        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
                        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        at org.grails.core.DefaultGrailsControllerClass$ReflectionInvoker.invoke(DefaultGrailsControllerClass.java:210)
                        at org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:187)
                        at org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:108)
                        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089)
                        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979)
                        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
                        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:903)
                        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885)
                        at org.springframework.security.web.FilterChainProxy.lambda$doFilterInternal$3(FilterChainProxy.java:231)
                        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:365)
                        at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:117)
                        at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:83)
                        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374)
                        at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:125)
                        at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:119)
                        at grails.plugin.springsecurity.web.UpdateRequestContextHolderExceptionTranslationFilter.doFilter(UpdateRequestContextHolderExceptionTranslationFilter.groovy:68)
                        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374)
                        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374)
                        at grails.plugin.springsecurity.web.filter.GrailsAnonymousAuthenticationFilter.doFilter(GrailsAnonymousAuthenticationFilter.groovy:56)
                        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374)
                        at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:150)
                        at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:105)
                        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374)
                        at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:179)
                        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374)
                        at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:235)
                        at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:229)
                        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374)
                        at grails.plugin.springsecurity.web.authentication.logout.MutableLogoutFilter.doFilter(MutableLogoutFilter.groovy:66)
                        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374)
                        at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:117)
                        at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:87)
                        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374)
                        at grails.plugin.springsecurity.web.SecurityRequestHolderFilter.doFilter(SecurityRequestHolderFilter.groovy:62)
                        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:374)
                        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:233)
                        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:191)
                        at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:80)
                        at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:70)
                        at java.base/java.lang.Thread.run(Thread.java:840)
                Caused by: org.codehaus.groovy.runtime.typehandling.GroovyCastException: Cannot cast object 'George Doe' with class 'java.lang.String' to class 'functionaltests.Person'
                        at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
                        ... 42 common frames omitted

--- a/grails-controllers/src/main/groovy/org/grails/compiler/web/ControllerActionTransformer.java
+++ b/grails-controllers/src/main/groovy/org/grails/compiler/web/ControllerActionTransformer.java
@@ -876,17 +876,11 @@ public class ControllerActionTransformer implements GrailsArtefactClassInjector,
 
         final ClassNode paramTypeClassNode = param.getType();
         final String methodParamName = param.getName();
-        Expression getParamsExpression = buildGetPropertyExpression(new VariableExpression("this"), "params", classNode);
-        final Expression paramsContainsKeyMethodArguments = new ArgumentListExpression(
-                new ConstantExpression(requestParameterName));
-        final BooleanExpression containsKeyExpression = new BooleanExpression(
-                applyDefaultMethodTarget(new MethodCallExpression(getParamsExpression, "containsKey", paramsContainsKeyMethodArguments), Map.class));
-        final Statement initializeParameterStatement = new ExpressionStatement(
-                new DeclarationExpression(new VariableExpression(
-                        methodParamName, paramTypeClassNode),
-                        Token.newSymbol(Types.EQUALS, 0, 0),
-                        new TernaryExpression(containsKeyExpression, buildGetMapExpression(getParamsExpression, requestParameterName) , new ConstantExpression(null))));
-        wrapper.addStatement(initializeParameterStatement);
+        Expression getParamsExpression = buildGetPropertyExpression(localVarX("this", classNode), "params", classNode);
+        final BooleanExpression containsKeyExpression = boolX(applyDefaultMethodTarget(callX(getParamsExpression, "containsKey", args(constX(requestParameterName))), Map.class));
+
+        wrapper.addStatement(stmt(declX(localVarX(methodParamName, paramTypeClassNode),
+                ternaryX(containsKeyExpression, buildGetMapExpression(getParamsExpression, requestParameterName), nullX()))));
     }
 
     protected void initializePrimitiveOrTypeWrapperParameter(final ClassNode classNode, final BlockStatement wrapper,

--- a/grails-test-examples/app1/src/integration-test/groovy/functionaltests/commandobjects/CommandObjectSpec.groovy
+++ b/grails-test-examples/app1/src/integration-test/groovy/functionaltests/commandobjects/CommandObjectSpec.groovy
@@ -22,13 +22,11 @@ package functionaltests.commandobjects
 import grails.plugin.geb.ContainerGebSpec
 import grails.testing.mixin.integration.Integration
 import spock.lang.Issue
-import spock.lang.PendingFeature
 
 @Integration
 @Issue('https://github.com/apache/grails-core/issues/14947')
 class CommandObjectSpec extends ContainerGebSpec {
 
-    @PendingFeature
     void 'should display the correct title on the home page'() {
         when: 'visiting the home page'
         go('/commandObject/echo?person=George+Doe')

